<launch>
	<arg name="n" default="1"/>
	<arg name="use_joystick" default="false"/>
	<arg name="use_sim_time" default="true"/>

	<arg name="robot_name" value="$(eval 'nesfr3_' + str(arg('n')))"/>
	<!-- <arg name="coordinate" value="$(eval arg('n') - 1)"/>  -->
	<arg name="coordinate" value="0.0"/>
	<arg name="namespace" value="$(eval 'nesfr3/' + str(arg('n')))"/>

	<group ns="$(arg namespace)">

		<group unless="$(arg use_joystick)">
			<node pkg="nesfr3_services" name="shot_controller_node" type="shot_controller" output="screen">
				<param name="robot_id" type="int" value="$(arg n)" />
				<param name="use_sim_time" value="$(arg use_sim_time)"/>
			</node>
		</group> 


		<param name="robot_description"
	 	command="$(find xacro)/xacro '$(find nesfr3_description)/urdf/nesfr3.xacro' index:=$(arg n)"  /> 

	 	<node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"/>

  		<node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
  			<param name="tf_prefix" value="$(arg namespace)" type="str"/>
  			<param name="use_sim_time" value="$(arg use_sim_time)"/>
  		</node> 
	 	
		<!-- this is for the lua file, might cause a race condition	 -->
	 	<env name="robot_index" value="$(arg n)"/>

	 	<node name="cartographer_node" pkg="cartographer_ros"
			type="cartographer_node" args="
			-configuration_directory $(find nesfr3_gazebo)/config
			-configuration_basename ouster_2d.lua"
			output="screen">
			
			<remap from="odom" to="wheel_odom" />
			<remap from="imu" to="lidar/imu" />
			<remap from="points2" to="lidar/points" />
		</node>


		<node name="cartographer_occupancy_grid_node" pkg="cartographer_ros"
			type="cartographer_occupancy_grid_node" args="-resolution 0.05" />


		<node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
			args="-urdf -model $(arg robot_name) -x -3 -y $(arg coordinate) -z $(arg coordinate) -param robot_description"/>
	</group>

	<include file="$(find nesfr3_gazebo)/launch/robots.launch" if="$(eval arg('n') - 1 > 0)">
       <arg name="n" value="$(eval arg('n') - 1)"/>
       <arg name="use_joystick" value="false"/>
       <arg name="use_sim_time" value="$(arg use_sim_time)"/>
   	</include>

</launch>